src/main/java/org/junitpioneer/jupiter/SystemPropertyExtension.java

```diff
@@ -0,0 +1,86 @@
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import java.lang.reflect.Method;
+import java.util.Properties;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.platform.commons.support.AnnotationSupport;
+
+class SystemPropertyExtension implements BeforeEachCallback, AfterEachCallback {
+
+	private static final Namespace NAMESPACE = Namespace.create(SystemPropertyExtension.class);
+
+	private static final String KEY = "SystemProperty";
+
+	@Override
+	public void beforeEach(final ExtensionContext context) throws Exception {
+		if (annotationsPresentOnTestMethod(context)) {
+			handleSystemProperties(context);
+		}
+	}
+
+	private boolean annotationsPresentOnTestMethod(ExtensionContext context) {
+		//@formatter:off
+		return context.getTestMethod()
+				.map(testMethod -> AnnotationSupport.isAnnotated(testMethod, ClearSystemProperty.class)
+						|| AnnotationSupport.isAnnotated(testMethod, ClearSystemProperties.class)
+						|| AnnotationSupport.isAnnotated(testMethod, SetSystemProperty.class)
+						|| AnnotationSupport.isAnnotated(testMethod, SetSystemProperties.class))
+				.orElse(false);
+		//@formatter:on
+	}
+
+	private void handleSystemProperties(final ExtensionContext context) {
+		storeOriginalSystemProperties(context);
+		context.getTestMethod().ifPresent(testMethod -> {
+			clearAnnotatedSystemProperties(testMethod);
+			setAnnotatedSystemProperties(testMethod);
+		});
+	}
+
+	private void storeOriginalSystemProperties(final ExtensionContext context) {
+		final Properties backup = new Properties();
+		backup.putAll(System.getProperties());
+		context.getStore(NAMESPACE).put(KEY, backup);
+	}
+
+	private void clearAnnotatedSystemProperties(final Method testMethod) {
+		//@formatter:off
+		AnnotationSupport.findRepeatableAnnotations(testMethod, ClearSystemProperty.class).stream()
+				.forEach(prop -> System.clearProperty(prop.key()));
+		//@formatter:on
+	}
+
+	private void setAnnotatedSystemProperties(final Method testMethod) {
+		//@formatter:off
+		AnnotationSupport.findRepeatableAnnotations(testMethod, SetSystemProperty.class).stream()
+				.forEach(prop -> System.setProperty(prop.key(), prop.value()));
+		//@formatter:on
+	}
+
+	@Override
+	public void afterEach(final ExtensionContext context) throws Exception {
+		if (annotationsPresentOnTestMethod(context)) {
+			resetOriginalSystemProperties(context);
+		}
+	}
+
+	private void resetOriginalSystemProperties(final ExtensionContext context) {
+		final Properties backup = context.getStore(NAMESPACE).get(KEY, Properties.class);
+		System.setProperties(backup);
```

Comment:
This restores all properties to the values they had before the test was run. If tests change system properties on purpose (sounds fishy, I know, but weirder things have happened), this extension overrides/negates that change. I think it would be better to just store/restore the properties used by the annotation.

To go further, maybe we should only restore system properties that have the value given to this extension? If a test sets any system property (even the one set for the test by this extension), we should not reset that. At least that's what my gut says atm. What do you think?

Reply:
I now only store and restore those system properties that have been specified via the extension, not all per se. But I think it is fine to restore a system property in this situation:

>If a test sets any system property (even the one set for the test by this extension), we should not reset that.

I would argue that this is defined behavior as specified in the Javadoc. I also believe other extensions such as `DefaultLocaleExtension` and `DefaultTimeZoneExtension` work the same.

What do you say?

Reply:
Makes sense, let's go with that. For now - maybe I'll come back to it in a separate issue.